File: app/src/main/java/com/fed/notes/view/NotePreviewFragment.java
package com.fed.notes.view;

import android.content.Intent;
import android.graphics.Bitmap;
import android.net.Uri;
import android.os.Build;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.design.widget.AppBarLayout;
import android.support.v4.app.Fragment;
import android.support.v4.app.FragmentManager;
import android.support.v4.app.ShareCompat;
import android.support.v4.content.FileProvider;
import android.support.v7.app.AlertDialog;
import android.support.v7.widget.Toolbar;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import com.fed.notes.App;
import com.fed.notes.BuildConfig;
import com.fed.notes.R;
import com.fed.notes.database.DbHelper;
import com.fed.notes.database.Note;
import com.fed.notes.utils.ImageDialog;
import com.fed.notes.utils.NotesOrderUtil;
import com.fed.notes.utils.PictureUtils;
import com.getbase.floatingactionbutton.FloatingActionButton;
import com.getbase.floatingactionbutton.FloatingActionsMenu;

import java.io.File;
import java.text.DateFormat;
import java.text.SimpleDateFormat;
import java.util.Locale;
import java.util.UUID;

import javax.inject.Inject;

import io.reactivex.schedulers.Schedulers;

/**
 * Created by f on 05.05.2017.
 */

public class NotePreviewFragment extends Fragment {
    private static final String ARGS_NOTE_ID = "argsnoteid";

    private Note note;
    private UUID noteID;

    private TextView noteTitleField;
    private TextView noteDescriptionField;
    private TextView date;
    private ImageView photoView;
    private AppBarLayout appBarLayout;
    private Toolbar toolbar;

    private FloatingActionsMenu fabMenu;
    private File photoFile;
    private Uri uriPhotoFile;

    @Inject
    DbHelper dbHelper;

    public static NotePreviewFragment newInstance(UUID id) {
        Bundle args = new Bundle();
        args.putSerializable(ARGS_NOTE_ID, id);

        NotePreviewFragment noteFragment = new NotePreviewFragment();
        noteFragment.setArguments(args);
        return noteFragment;
    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        App.getComponent().inject(this);

        noteID = (UUID) getArguments().getSerializable(ARGS_NOTE_ID);
//        dbHelper.getNoteRx(noteID)
//                .subscribeOn(Schedulers.io())
//                .observeOn(AndroidSchedulers.mainThread())
//                .subscribe(
//                    nt -> {
//                        note = nt;
//                        photoFile = dbHelper.getPhotoFile(note);
//                        if (Build.VERSION.SDK_INT > 23) {
//                            uriPhotoFile = FileProvider.getUriForFile(getActivity(), BuildConfig.APPLICATION_ID + ".provider", photoFile);
//                        } else {
//                            uriPhotoFile = Uri.fromFile(photoFile);
//                        }
//                    },
//                    Throwable::printStackTrace
//                );
    }

    @Nullable
    @Override
    public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container, Bundle savedInstanceState) {
        View v = inflater.inflate(R.layout.fragment_note_preview, container, false);

        toolbar = v.findViewById(R.id.toolbar);
        noteTitleField = v.findViewById(R.id.note_title);
        noteDescriptionField = v.findViewById(R.id.note_description);
        date = v.findViewById(R.id.create_date);
        appBarLayout = v.findViewById(R.id.appbar);
        photoView = v.findViewById(R.id.note_photo);
        photoView.setOnClickListener(v1 -> {
            FragmentManager manager = getFragmentManager();
            ImageDialog dialog = ImageDialog.newInstance(photoFile.getPath());
            dialog.show(manager, "IMAGE_FULL");
        });

        note = dbHelper.getNote(noteID);
        if (note != null) {
            initFabs(v);
            updateInfo();
            updatePhotoView();
        }
        return v;
    }

    private void initFabs(View v) {
        fabMenu = v.findViewById(R.id.fab_menu);

        FloatingActionButton fabSend = v.findViewById(R.id.fab_send);
        fabSend.setIcon(R.drawable.ic_send_email);
        fabSend.setOnClickListener(view -> sendEmailDialog());

        FloatingActionButton fabEdit = v.findViewById(R.id.fab_edit);
        fabEdit.setIcon(R.drawable.ic_edit_mode);
        fabEdit.setOnClickListener(view -> editNoteFragment());

        FloatingActionButton fabDel = v.findViewById(R.id.fab_delete);
        fabDel.setIcon(R.drawable.ic_delete_note);
        fabDel.setOnClickListener(view -> deleteNoteDialog());
    }

    private void updateInfo() {
        photoFile = dbHelper.getPhotoFile(note);

        if (Build.VERSION.SDK_INT > 23) {
            uriPhotoFile = FileProvider.getUriForFile(getActivity(), BuildConfig.APPLICATION_ID + ".provider", photoFile);
        } else {
            uriPhotoFile = Uri.fromFile(photoFile);
        }

        toolbar.setTitle(note.title);
        if (note.title != null) {
            noteTitleField.setVisibility(View.VISIBLE);
            noteTitleField.setText(note.title);
        } else {
            noteTitleField.setVisibility(View.GONE);
        }
        if (note.description != null) {
            noteDescriptionField.setVisibility(View.VISIBLE);
            noteDescriptionField.setText(note.description);
        } else {
            noteDescriptionField.setVisibility(View.GONE);
        }
        DateFormat dateFormat = new SimpleDateFormat(getString(R.string.date_format), Locale.ENGLISH);
        date.setText(dateFormat.format(note.date));
    }

    private void updatePhotoView() {
        if (photoFile.exists()) {
            Bitmap bitmap = PictureUtils.getScaledBitmap(photoFile.getPath(), getActivity());
            photoView.setImageBitmap(bitmap);
            appBarLayout.setVisibility(View.VISIBLE);
            noteTitleField.setVisibility(View.GONE);
        } else {
            appBarLayout.setVisibility(View.GONE);
            noteTitleField.setVisibility(View.VISIBLE);
        }
//        photoView.setImageURI(Uri.fromFile(photoFile));
    }

    private void sendEmailDialog() {
        Intent intent;
        if (photoFile.exists()) {
            intent = ShareCompat.IntentBuilder.from(getActivity())
                    .setType("plain/text")
//                            .setType("image/*")
                    .setSubject(getResources().getString(R.string.email_text) + note.title)
                    .setText(note.description)
                    .setStream(uriPhotoFile)
                    .getIntent();
        } else {
            intent = ShareCompat.IntentBuilder.from(getActivity())
                    .setType("plain/text")
                    .setSubject(getResources().getString(R.string.email_text) + note.title)
                    .setText(note.description)
                    .getIntent();
        }
        if (intent.resolveActivity(getActivity().getPackageManager()) != null) {
            startActivity(intent);
//                    startActivity(Intent.createChooser(intent, "send via..."));
        } else {
            AlertDialog.Builder eMailIntentAlertDialog = new AlertDialog.Builder(getActivity());
            eMailIntentAlertDialog.setTitle(R.string.email_intent_title);
            eMailIntentAlertDialog.setMessage(R.string.email_intent_text);
            eMailIntentAlertDialog.setNeutralButton(R.string.ok_button, (dialog, which) -> {
            });
            eMailIntentAlertDialog.show();
        }
    }

    private void editNoteFragment() {
        ((MainActivity) getActivity()).openNoteFragmentEditor(note);
    }

    private void deleteNoteDialog() {
        AlertDialog.Builder deleteAlertDialog = new AlertDialog.Builder(getActivity());
        deleteAlertDialog.setTitle(R.string.alert_on_del_title);
        deleteAlertDialog.setMessage(R.string.alert_on_del_text);
        deleteAlertDialog.setPositiveButton(R.string.alert_on_del_yes, (dialog, which) -> {
            NotesOrderUtil.removeNoteFromOrderList(note.id, getContext());
            dbHelper.deleteRx(note)
                    .observeOn(Schedulers.io())
                    .subscribe(() -> {
                            },
                            Throwable::printStackTrace);
            NotePreviewFragment.this.getActivity().onBackPressed();
        });
        deleteAlertDialog.setNeutralButton(R.string.alert_on_del_cancel, (dialog, which) -> {
        });
        deleteAlertDialog.show();
    }

    @Override
    public void onResume() {
        super.onResume();
        fabMenu.collapse();
        if (note == null) {
            getActivity().getSupportFragmentManager().popBackStack();
        }
    }
}

File: app/src/main/java/com/fed/notes/view/NotePreviewFragment.kt
package com.fed.notes.view

import android.content.Intent
import android.os.Bundle
import android.support.v4.app.Fragment
import android.support.v4.app.ShareCompat
import android.support.v7.app.AlertDialog
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import com.fed.notes.App
import com.fed.notes.R
import com.fed.notes.database.DbHelper
import com.fed.notes.database.Note
import com.fed.notes.utils.*
import io.reactivex.schedulers.Schedulers
import kotlinx.android.synthetic.main.fragment_note_preview.*
import java.io.File
import java.util.*
import javax.inject.Inject

class NotePreviewFragment : Fragment() {
    companion object {
        private val ARGS_NOTE_ID = "args_note_id"

        fun newInstance(id: UUID): NotePreviewFragment {
            val args = Bundle()
            args.putSerializable(ARGS_NOTE_ID, id)

            val noteFragment = NotePreviewFragment()
            noteFragment.arguments = args
            return noteFragment
        }
    }

    @Inject
    lateinit var dbHelper: DbHelper

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        App.component?.inject(this)
    }

    override fun onCreateView(inflater: LayoutInflater?, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        return container?.inflate(R.layout.fragment_note_preview)
    }

    override fun onViewCreated(view: View?, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        val noteID = arguments.getSerializable(ARGS_NOTE_ID) as UUID

        val note = dbHelper.getNote(noteID)
        if (note == null) {
            activity.supportFragmentManager.popBackStack()
        } else {
            showNote(note)
        }

//        dbHelper.getNoteRx(noteID)
//                .subscribeOn(Schedulers.io())
//                .observeOn(AndroidSchedulers.mainThread())
//                .subscribe({
//                    note = it
//                    showNote()
//                }, Throwable::printStackTrace)
    }

    override fun onResume() {
        super.onResume()
        fab_menu.collapse()
    }

    private fun showNote(note: Note) {
        val photoFile = dbHelper.getPhotoFile(note)

        showTextFromNote(note)
        initFabs(photoFile, note)
        updatePhotoView(photoFile)
    }

    private fun initFabs(photoFile: File?, note: Note) {
        fab_send.setIcon(R.drawable.ic_send_email)
        fab_send.setOnClickListener { onSendEmailClicked(photoFile, note) }

        fab_edit.setIcon(R.drawable.ic_edit_mode)
        fab_edit.setOnClickListener { editNoteFragment(note) }

        fab_delete.setIcon(R.drawable.ic_delete_note)
        fab_delete.setOnClickListener { deleteNoteDialog(note) }
    }

    private fun showTextFromNote(note: Note) {
        toolbar.title = note.title

        note.title?.let {
            note_title_text_view.visibility = View.VISIBLE
            note_title_text_view.text = note.title
        }

        note.description?.let {
            note_description_text_view.visibility = View.VISIBLE
            note_description_text_view.text = note.description
        }

        note_date_text_view.showDate(context, note.date)
    }

    private fun updatePhotoView(photoFile: File?) {
        if (photoFile!!.exists()) {
            val bitmap = PictureUtils.getScaledBitmap(photoFile.path, activity)
            note_photo_image_view.setImageBitmap(bitmap)
            note_photo_image_view.setOnClickListener { showPhotoDialog(photoFile.path) }
            appbar_layout.visibility = View.VISIBLE
            note_title_text_view.visibility = View.GONE
        } else {
            appbar_layout.visibility = View.GONE
            note_title_text_view.visibility = View.VISIBLE
        }
    }

    private fun showPhotoDialog(path: String) {
        val dialog = ImageDialog.newInstance(path)
        dialog.show(fragmentManager, "IMAGE_FULL")
    }

    private fun onSendEmailClicked(photoFile: File?, note: Note) {

        val intent: Intent = if (photoFile?.exists() == true) {
            getImageIntent(photoFile, note)
        } else {
            getTextIntent(note)
        }

        if (intent.resolveActivity(activity.packageManager) != null) {
            startActivity(intent)
        } else {
            showEmailDialog()
        }
    }

    private fun getImageIntent(photoFile: File?, note: Note): Intent {
        val uriPhotoFile = getUri(context, photoFile)
        return ShareCompat.IntentBuilder.from(activity)
                .setType("plain/text")
                .setSubject(resources.getString(R.string.email_text) + note.title)
                .setText(note.description)
                .setStream(uriPhotoFile)
                .intent
    }

    private fun getTextIntent(note: Note): Intent {
        return ShareCompat.IntentBuilder.from(activity)
                .setType("plain/text")
                .setSubject(resources.getString(R.string.email_text) + note.title)
                .setText(note.description)
                .intent
    }

    private fun showEmailDialog() {
        AlertDialog.Builder(activity).apply {
            setTitle(R.string.email_intent_title)
            setMessage(R.string.email_intent_text)
            setNeutralButton(R.string.ok_button, null)
        }.show()
    }

    private fun editNoteFragment(note: Note) = (activity as MainActivity).openNoteFragmentEditor(note)

    private fun deleteNoteDialog(note: Note) {
        AlertDialog.Builder(activity).apply {
            setTitle(R.string.alert_on_del_title)
            setMessage(R.string.alert_on_del_text)
            setPositiveButton(R.string.alert_on_del_yes) { _, _ -> onDeleteClicked(note) }
            setNeutralButton(R.string.alert_on_del_cancel, null)
        }.show()
    }

    private fun onDeleteClicked(note: Note) {
        removeNoteFromOrderList(note.id, context)
        dbHelper.deleteRx(note)
                .subscribeOn(Schedulers.io())
                .subscribe({},
                        Throwable::printStackTrace)
        activity.onBackPressed()
    }
}